Tweak DiagnosticEngine's 'aka' logic to only kick in for typealiases.

jrose-apple opened the pull request:
Previously we had more ad hoc logic that tried to decide if it was worth desugaring a type based on its structure. Now we instead look for a typealias that might actually benefit from desugaring, and if we don't find one we won't show the 'aka' note.

(This will be used for #8987.)

Some of the output involving function types gets a lot wordier from this, but I'm not convinced that's the wrong thing to do. The mismatch might be within a type alias!
rdar://problem/31511958

jrose-apple commented:
@swift-ci Please test

slavapestov commented on lib/AST/DiagnosticEngine.cpp:
```diff
@@ -313,6 +314,42 @@ static void formatSelectionArgument(StringRef ModifierArguments,
   
 }
 
+static bool isInterestingTypealias(Type type) {
```
Great, this is exactly what I need for 'AnyObject' as well.

jrose-apple commented:
@swift-ci Please test Linux

jrose-apple committed:
Tweak DiagnosticEngine's 'aka' logic to only kick in for typealiases.

Previously we had more ad hoc logic that tried to decide if it was
worth desugaring a type based on its structure. Now we instead look
for a typealias that might actually benefit from desugaring, and if
we don't find one we won't show the 'aka' note.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test macOS
